pkg/sql/executor.go

```diff
@@ -1604,45 +1604,50 @@ func (e *Executor) execStmtInOpenTxn(
-			if parErr := session.synchronizeParallelStmts(session.Ctx()); parErr != nil {
-				// If synchronizing results in a non-retryable error, it takes priority.
-				if _, ok := parErr.(*roachpb.HandledRetryableTxnError); !ok {
-					err = parErr
-					return
+		// If we can still automatically retry the txn, and we detect that the
+		// transaction won't be allowed to commit because its timestamp has been
+		// pushed, we go ahead and retry now - if we'd execute further statements,
+		// we probably wouldn't be allowed to retry automatically any more.
+		// We want to propagate the error if either:
+		// 1.we're in AutoRetry, for the reason described above.
+		// 2.retryIntent is not set, because the client would not benefit from letting the transaction run more statements.
+		if txnState.isSerializableRestart() {
+			if txnState.State() == AutoRetry || !txnState.retryIntent {
+				// If we just ran a statement synchronously, then the parallel queue
+				// would have been synchronized first, so this would be a no-op.
+				// However, if we just ran a statement asynchronously, then the
+				// queue could still contain statements executing. So if we're in a
+				// SerializableRestart state, we synchronize to drain the rest of
+				// the stmts and clear the parallel batch's error-set before
+				// restarting. If we did not drain the parallel stmts then the
+				// txn.Reset call below might cause them to write at the wrong
+				// epoch.
+				//
+				// TODO(nvanbenschoten): like in Session.Finish, we should try to
+				// actively cancel these parallel queries instead of just waiting
+				// for them to finish.
+				if parErr := session.synchronizeParallelStmts(session.Ctx()); parErr != nil {
+					// If synchronizing results in a non-retryable error, it takes priority.
+					if _, ok := parErr.(*roachpb.HandledRetryableTxnError); !ok {
+						err = parErr
+						return
+					}
 				}
 			}
-
-			// If we can still automatically retry the txn, and we detect that the
-			// transaction won't be allowed to commit because its timestamp has been
-			// pushed, we go ahead and retry now - if we'd execute further statements,
-			// we probably wouldn't be allowed to retry automatically any more.
 			txnState.mu.txn.Proto().Restart(
 				0 /* userPriority */, 0 /* upgradePriority */, e.cfg.Clock.Now())
-			// Force an auto-retry by returning a retryable error to the higher
-			// levels.
-			err = roachpb.NewHandledRetryableTxnError(
-				"serializable transaction timestamp pushed (detected by SQL Executor)",
-				txnState.mu.txn.ID(),
-				// No updated transaction required; we've already manually updated our
-				// client.Txn.
-				roachpb.Transaction{},
-			)
-			err = txnState.updateStateAndCleanupOnErr(err, e)
+			// if we already knew the serializable transaction has been pushed by
+			// another transaction, we should return an retriable error to the higher level.
+			if txnState.State() == AutoRetry || !txnState.retryIntent {
```

Comment:
looks like you've tried to address my comment about only returning the error sometimes, but the code is still wrong. You also don't want to run the `txnState.mu.txn.Proto().Restart()` above unless we're going to return this error.
The whole big block that starts with `if txnState.isSerializableRestart()` should start with `if txnState.isSerializableRestart() && (AutoRetry || !retryIntent)`.

Reply:
Sorry, I just couldn't understand why we should run the `txnState.mu.txn.Proto().Restart()` when `txnState.isSerializableRestart() && !retryIntent`?